src/hotspot/share/nmt/virtualMemoryTracker.cpp

```diff
@@ -47,545 +46,213 @@ void VirtualMemory::update_peak(size_t size) {
 
 void VirtualMemorySummary::snapshot(VirtualMemorySnapshot* s) {
   // Snapshot current thread stacks
-  VirtualMemoryTracker::snapshot_thread_stacks();
+  VirtualMemoryTracker::Instance::snapshot_thread_stacks();
   as_snapshot()->copy_to(s);
 }
 
-SortedLinkedList<ReservedMemoryRegion, compare_reserved_region_base>* VirtualMemoryTracker::_reserved_regions;
-
-int compare_committed_region(const CommittedMemoryRegion& r1, const CommittedMemoryRegion& r2) {
-  return r1.compare(r2);
+bool VirtualMemoryTracker::Instance::initialize(NMT_TrackingLevel level) {
+  assert(_tracker == nullptr, "only call once");
+  if (level >= NMT_summary) {
+    _tracker = static_cast<VirtualMemoryTracker*>(os::malloc(sizeof(VirtualMemoryTracker), mtNMT));
+    if (_tracker == nullptr) return false;
+    new (_tracker) VirtualMemoryTracker(level == NMT_detail);
+    return _tracker->tree() != nullptr;
```

Comment:
We should check for `tree() != nullptr;` inside VirtualMemoryTracker constructor as assert?

Reply:
Then in Product build, the `tree == nullptr` ends up in a crash.